### DISCUSSION TITLE
Automatic IDs

### DISCUSSION THREAD

**Discussion by innocenzi** [chakra-ui/zag#905] [Ideas]
# 🚀 Feature request

I am wondering if it'd be possible to automatically generate an ID, instead of requiring the user to specify one.

## 🧱 Problem Statement / Justification

Disclaimer; I haven't used Zag yet. I discovered it a while ago, and only now reading the docs. I like the concept a lot, but reading the docs, I keep seeing the requirement to provide an ID.

I think it would be better if that was optional, so the developer doesn't have to manage it.

## ✅ Proposed solution or API

In Headless UI, they use a simple [`useId`](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-vue/src/hooks/use-id.ts) hook internally. Here's [an example](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-vue/src/components/disclosure/disclosure.ts#L78) of how they use it.

**Comment by segunadebayo** [comment]
Hi @innocenzi,

Zag.js is largely framework agnostic, and there's no way we can generate unique IDs that work in SSR environments. That's why we leave it to the userland.

The idea you shared broadly inspired us to work on [Ark UI](https://ark-ui.com/), a sister project that consumes Zag.js and provides headless components that do all the heavy lifting for the users by setting a unique ID out of the box.

**Comment by segunadebayo** [comment]
Closing — IDs are now optional. Machines generate stable DOM IDs automatically.
